src/source/from-shift.js

```diff
@@ -0,0 +1,205 @@
+'use strict';
+
+function makeEager(obj) {
+	obj.type = 'Eager' + obj.type;
+}
+
+function dummyDeclaredScope(type) {
+	return {
+		type,
+		declaredNames: [],
+		hasDirectEval: false
+	};
+}
+
+function parameterScopeAndLength(params) {
+	let length = 0;
+	let isSimpleParameterList = true;
+
+	loop: for (let param of params) {
+		switch (param.type) {
+			case 'BindingIdentifier':
+				length++;
+				break;
+			case 'ObjectBinding':
+			case 'ArrayBinding':
+				isSimpleParameterList = false;
+				length++;
+				break;
+			default:
+				isSimpleParameterList = false;
+				break loop;
+		}
+	}
+
+	return {
+		parameterScope: {
+			type: 'AssertedParameterScope',
+			paramNames: [],
+			hasDirectEval: false,
+			isSimpleParameterList
+		},
+		length
+	};
+}
+
+function dummyBoundNamesScope() {
+	return {
+		type: 'AssertedBoundNamesScope',
+		boundNames: [],
+		hasDirectEval: true
+	};
+}
+
+function createFunctionContents(obj) {
+	let kind = obj.type;
+
+	let directives = obj.body.directives || [];
+	delete obj.body.directives;
+
+	let isExpressionBody;
+	let body = obj.body;
+	delete obj.body;
+
+	if (body.type === 'FunctionBody') {
+		body = body.statements;
+		isExpressionBody = false;
+	} else {
+		isExpressionBody = true;
+	}
+
+	let contents;
+
+	switch (kind) {
+		case 'FunctionDeclaration':
+		case 'Method':
+			contents = {
+				type: 'FunctionOrMethodContents',
+				isThisCaptured: false
+			};
+			break;
+		case 'FunctionExpression':
+			contents = {
+				type: 'FunctionExpressionContents',
+				isFunctionNameCaptured: false,
+				isThisCaptured: false
+			};
+			break;
+		case 'ArrowExpression':
+			if (isExpressionBody) {
+				contents = { type: 'ArrowExpressionContentsWithExpression' };
+				obj.type = 'ArrowExpressionWithExpression';
+			} else {
+				contents = { type: 'ArrowExpressionContentsWithFunctionBody' };
+				obj.type = 'ArrowExpressionWithFunctionBody';
+			}
+			break;
+		case 'Getter':
+			contents = {
+				type: 'GetterContents',
+				isThisCaptured: false
+			};
+			break;
+		case 'Setter': {
+			let { param } = obj;
+			delete obj.param;
+			let { parameterScope, length } = parameterScopeAndLength([param]);
+			contents = {
+				type: 'SetterContents',
+				isThisCaptured: false,
+				param,
+				parameterScope
+			};
+			obj.length = length;
+			break;
+		}
+		default:
+			throw new TypeError(`Unexpected function kind ${kind}`);
+	}
+
+	if (kind !== 'Getter' && kind !== 'Setter') {
+		let { params } = obj;
+		delete obj.params;
+
+		contents.params = params;
+
+		let { parameterScope, length } = parameterScopeAndLength(params.items);
+
+		contents.parameterScope = parameterScope;
+		obj.length = length;
+		obj.isAsync = !!obj.isAsync;
+	}
+
+	contents.bodyScope = dummyDeclaredScope('AssertedVarScope');
+	contents.body = body;
+
+	obj.contents = contents;
+	obj.directives = directives;
+
+	makeEager(obj);
+}
+
+function convertObject(obj) {
```

Comment:
That's a callback for `JSON.stringify`, right? Can you mention it somewhere?

Reply:
No, that one is below.

Reply:
Added comment to the corresponding function.